modules/core/src/main/java/org/apache/ignite/internal/cdc/IgniteCDC.java

```diff
@@ -0,0 +1,485 @@
+                log.info("CDC dir     -\t" + cdcDir);
+                log.info("Binary meta -\t" + binaryMeta);
+                log.info("Marshaller  -\t" + marshaller);
+                log.info("--------------------------------");
+            }
+
+            state = new CDCConsumerState(cdcDir.resolve(STATE_DIR), "cdc-state.bin");
+
+            initState = state.load();
+
+            if (initState != null && log.isInfoEnabled())
+                log.info("Loaded initial state[state=" + initState + ']');
+
+            consumer.start(cfg, log);
+
+            try {
+                Predicate<Path> walFilesOnly = p -> WAL_NAME_PATTERN.matcher(p.getFileName().toString()).matches();
+
+                Comparator<Path> sortByNumber = Comparator.comparingLong(this::segmentIndex);
+
+                long[] lastSgmnt = new long[] { -1 };
+
+                waitFor(cdcDir, walFilesOnly, sortByNumber, segment -> {
+                    try {
+                        long nextSgmnt = segmentIndex(segment);
+
+                        assert lastSgmnt[0] == -1 || nextSgmnt - lastSgmnt[0] == 1;
+
+                        lastSgmnt[0] = nextSgmnt;
+
+                        readSegment(segment);
+
+                        return true;
+                    }
+                    catch (IgniteCheckedException | IOException e) {
+                        throw new IgniteException(e);
+                    }
+                }, cdcCfg.getSleepBeforeCheckNewSegmentsTimeout(), log);
+            }
+            finally {
+                consumer.stop();
+
+                if (log.isInfoEnabled())
+                    log.info("Ignite CDC Application stoped.");
+            }
+        }
+    }
+
+    /** Searches required directories. */
+    private void init() throws IOException {
+        String workDir = workDir(cfg);
+        String consIdDir = cdcDir.getName(cdcDir.getNameCount() - 1).toString();
+
+        Files.createDirectories(cdcDir.resolve(STATE_DIR));
+
+        binaryMeta = CacheObjectBinaryProcessorImpl.binaryWorkDir(workDir, consIdDir);
+
+        marshaller = MarshallerContextImpl.mappingFileStoreWorkDir(workDir);
+
+        if (log.isDebugEnabled()) {
+            log.debug("Using BinaryMeta directory[dir=" + binaryMeta + ']');
+            log.debug("Using Marshaller directory[dir=" + marshaller + ']');
+        }
+    }
+
+    /** Reads all available records from segment. */
+    private void readSegment(Path segment) throws IgniteCheckedException, IOException {
+        log.info("Processing WAL segment[segment=" + segment + ']');
+
+        IgniteWalIteratorFactory.IteratorParametersBuilder builder = new IgniteWalIteratorFactory.IteratorParametersBuilder()
+            .log(log)
+            .binaryMetadataFileStoreDir(binaryMeta)
+            .marshallerMappingFileStoreDir(marshaller)
+            .keepBinary(cdcCfg.isKeepBinary())
+            .filesOrDirs(segment.toFile())
+            .addFilter((type, ptr) -> type == DATA_RECORD_V2);
```

Comment:
The addition of a new type e.g. `DATA_RECORD_V3` should be safe for developers. If the type will change and we forgot to change this filter to the new type here the cdc would not work.

Reply:
`CDCSelfTest` will fail in that case.
What solution do you have in mind?